### ISSUE TITLE
"Overriding property must be as accessible as its enclosing type" while generating mock file

### ISSUE THREAD

**Issue by sanoj00b** [Brightify/Cuckoo#316]
I am using Xcode 11.3 and swift 5. 
Mac - 10.14.6 Mojave

I am using Cuckoo "1.1.1" and I start getting below error. I resolved the error using suggestion but once you build it will show again.

<img width="480" alt="Screenshot 2019-10-22 at 4 55 23 PM" src="https://user-images.githubusercontent.com/40563563/67281473-18ae1880-f4ed-11e9-9167-085fd571463c.png">

2- One more issue is:

**Value of type 'Class' has no member 'function':**  even though function has the internal access level. 

If anybody else facing same issue? please help.

**Comment by MatyasKriz** [comment]
Hey, can you please provide us with the mocked class so that we may try to reproduce this issue?

**Comment by sanoj00b** [comment]
@MatyasKriz 

```
override var empInfo: EmpInfo { # Error at this line: Overriding property must be as accessible as its enclosing type
        get {
            return cuckoo_manager.getter("empInfo",
                superclassCall:
                    
                    super.empInfo
                    ,
                defaultCall: __defaultImplStub!.empInfo)
        }
        
        set {
            cuckoo_manager.setter("empInfo",
                value: newValue,
                superclassCall:
                    
                    super.empInfo = newValue
                    ,
                defaultCall: __defaultImplStub!.empInfo = newValue)
        }
        
    }
```
Above is the mock property in the mocked class. 


```
open class EmpInfo: Model,Mappable {
    var empId: String!    
    open func mapping(map: Map) {
        empId <- map["empId"]
        super.mapping(map: map)
    }
}

```
This issue is still exist. I have updated my Xcode  to 11.2.1

**Comment by MatyasKriz** [comment]
Thanks, @sanoj00b. I'll try to reproduce the error and get back to you. Next time, please fork the project and add the reproducer to the tests. Way easier for us and much better chance we'll get to it faster.

**Comment by Peter-Nagy** [comment]
I did some investigation regarding this issue and I believe it's caused by `@testable` imports. When you want to mock something that's defined in another module and a function has `internal` visibility `@testable` makes it public, but when overriding it in the mocks Cuckoo doesn't add `public` accessibility to that function.

I'll try to fork the repo and reproduce it.

**Comment by Peter-Nagy** [comment]
Managed to reproduce the issue on this branch: https://github.com/Peter-Nagy/Cuckoo/tree/accessibility_bug_reproducing

**Comment by MatyasKriz** [comment]
Hey @Peter-Nagy, thanks for reproducing this problem. I'm currently pretty low on time, so I can't go fixing it at the moment. If you'd like to take a shot at it and create a PR, we'd greatly appreciate it.

**Comment by Peter-Nagy** [comment]
I'm happy to contribute but I'll need some guidance. 

In order to fix this problem, when importing something with `@testable` we'll have to bump up all overridden function/property accessibilities by one level. I'm not sure if it's possible at all - can we somehow find out if a class was imported with `@testable` or just with a normal import?

The other solution would be to simply bump up all accessibilities regardless of the module the class is imported from. Since these are just mock objects, I don't really see a big issue doing this, and it would be a much easier fix. We could also hide this behaviour behind a parameter, so the generator would only modify accessibilities if you explicitly request it.

**Comment by MatyasKriz** [comment]
I'm not sure what kind of beast we're dealing with here, so I'd like to hear what @TadeasKriz has to say on the matter. I was fixing some accessibility problems earlier and all it did was make problems in other use-cases that were working at the time.

**Comment by sanoj00b** [comment]
@MatyasKriz  any plan to merge the solution?

**Comment by sanoj00b** [comment]
with the latest version, I am facing the below issue.
```
** BUILD SUCCEEDED **

/usr/bin/xcrun xcodebuild -project /Users/***/Documents/Workspace/***/Carthage/Checkouts/Cuckoo/Cuckoo.xcodeproj -scheme Cuckoo+OCMock-iOS -configuration Release -derivedDataPath /Users/***/Library/Caches/org.carthage.CarthageKit/DerivedData/11.3_11C29/Cuckoo/1.2.0 -sdk iphoneos ONLY_ACTIVE_ARCH=NO CODE_SIGNING_REQUIRED=NO CODE_SIGN_IDENTITY= CARTHAGE=YES archive -archivePath /var/folders/82/_m5xwhxx4795wyxlhtw40ch509_hqq/T/Cuckoo SKIP_INSTALL=YES GCC_INSTRUMENT_PROGRAM_FLOW_ARCS=NO CLANG_ENABLE_CODE_COVERAGE=NO STRIP_INSTALLED_PRODUCT=NO (launched in /Users/***/Documents/Workspace/***/Carthage/Checkouts/Cuckoo)User defaults from command line:
    IDEArchivePathOverride = /var/folders/82/_m5xwhxx4795wyxlhtw40ch509_hqq/T/Cuckoo
    IDEDerivedDataPathOverride = /Users/***/Library/Caches/org.carthage.CarthageKit/DerivedData/11.3_11C29/Cuckoo/1.2.0

Build settings from command line:
    CARTHAGE = YES
    CLANG_ENABLE_CODE_COVERAGE = NO
    CODE_SIGN_IDENTITY = 
    CODE_SIGNING_REQUIRED = NO
    GCC_INSTRUMENT_PROGRAM_FLOW_ARCS = NO
    ONLY_ACTIVE_ARCH = NO
    SDKROOT = iphoneos13.2
    SKIP_INSTALL = YES
    STRIP_INSTALLED_PRODUCT = NO

note: Using new build system
note: Planning build
note: Constructing build description
error: /Users/***/Documents/Workspace/***/Carthage/Checkouts/Cuckoo/Pods/Target Support Files/Pods-Cuckoo+OCMock-iOS/Pods-Cuckoo+OCMock-iOS.release.xcconfig: unable to open file (in target "Cuckoo+OCMock-iOS" in project "Cuckoo") (in target 'Cuckoo+OCMock-iOS' from project 'Cuckoo')
error: /Users/***/Documents/Workspace/***/Carthage/Checkouts/Cuckoo/Pods/Target Support Files/Pods-Cuckoo+OCMock-iOS/Pods-Cuckoo+OCMock-iOS.release.xcconfig: unable to open file (in target "Cuckoo+OCMock-iOS" in project "Cuckoo") (in target 'Cuckoo+OCMock-iOS' from project 'Cuckoo')
error: /Users/***/Documents/Workspace/***/Carthage/Checkouts/Cuckoo/Pods/Target Support Files/Pods-Cuckoo+OCMock-iOS/Pods-Cuckoo+OCMock-iOS.release.xcconfig: unable to open file (in target "Cuckoo+OCMock-iOS" in project "Cuckoo") (in target 'Cuckoo+OCMock-iOS' from project 'Cuckoo')

** ARCHIVE FAILED **

```

any clue?

**Comment by sanoj00b** [comment]
> I'm happy to contribute but I'll need some guidance.
> 
> In order to fix this problem, when importing something with `@testable` we'll have to bump up all overridden function/property accessibilities by one level. I'm not sure if it's possible at all - can we somehow find out if a class was imported with `@testable` or just with a normal import?
> 
> The other solution would be to simply bump up all accessibilities regardless of the module the class is imported from. Since these are just mock objects, I don't really see a big issue doing this, and it would be a much easier fix. We could also hide this behaviour behind a parameter, so the generator would only modify accessibilities if you explicitly request it.

I am facing this issue. Could you please tell me specifically what needs to be done.

**Comment by sanoj00b** [comment]
> I'm happy to contribute but I'll need some guidance.
> 
> In order to fix this problem, when importing something with `@testable` we'll have to bump up all overridden function/property accessibilities by one level. I'm not sure if it's possible at all - can we somehow find out if a class was imported with `@testable` or just with a normal import?
> 
> The other solution would be to simply bump up all accessibilities regardless of the module the class is imported from. Since these are just mock objects, I don't really see a big issue doing this, and it would be a much easier fix. We could also hide this behaviour behind a parameter, so the generator would only modify accessibilities if you explicitly request it.

@Peter-Nagy  hey, just wondering if you fix the mentioned issue. Please explain it a bit more.

**Comment by TadeasKriz** [comment]
@Peter-Nagy Sorry for the delay, I had a lot on my plate. Anyway, a proper fix is incoming today.

**Comment by TadeasKriz** [comment]
@Peter-Nagy @sanoj00b Could you please try branch `fix/316` and let me know if it works for you?

**Comment by MatyasKriz** [comment]
I would like to thank both @Peter-Nagy for the research and reproduction of the issue and @TadeasKriz for solving this one. Swift accessibilities are a tad difficult for me to wrap my head around, so thanks again to both of you. 🙂

Let's hope that this is the final time I hear of accessibility issues. 😀

**Comment by sanoj00b** [comment]
> @Peter-Nagy @sanoj00b Could you please try branch `fix/316` and let me know if it works for you?

for me it's not working with the mentioned branch, still the same issue. Please let me know if it works for it others

**Comment by TadeasKriz** [comment]
@sanoj00b The changes were only in the Generator. Could you make sure you're generating the code with the generator compiled from the branch `fix/316`?

**Comment by sanoj00b** [comment]
> @sanoj00b The changes were only in the Generator. Could you make sure you're generating the code with the generator compiled from the branch `fix/316`?
Yes, branch name is the same. I used it and after running the carthage update command cartfile generating following commit id which is correct and I have verified this commit id.
github "SwiftKit/Cuckoo" "1d31b0634ccf37f6a3ef043254512192910c4050"

**Comment by sanoj00b** [comment]
> > @sanoj00b The changes were only in the Generator. Could you make sure you're generating the code with the generator compiled from the branch `fix/316`?

Yes, branch name is the same. I used it and after running the carthage update command cartfile generating following commit id which is correct and I have verified this commit id.
github "SwiftKit/Cuckoo" "1d31b0634ccf37f6a3ef043254512192910c4050"

**Comment by TadeasKriz** [comment]
Carthage downloads only the runtime framework, not the Generator binary. You probably have to do that manually.

**Comment by sanoj00b** [comment]
> Carthage downloads only the runtime framework, not the Generator binary. You probably have to do that manually.
I tried doing manually as well, but as soon as I build it again those file regenerated again, error comes again.

**Comment by sanoj00b** [comment]
@MatyasKriz  @TadeasKriz 
Still, I am facing the same issue. 
`Overriding property must be as accessible as its enclosing type Insert 'public '`
Even though If I add manually these 'public' specifier after building again, these same error comes again.

**Comment by Peter-Nagy** [comment]
I've merged the fix branch to my fork where the issue is reproduced, unfortunately it still isn't working for me.

I think you might've misunderstood the issue, so here is a quick example:

Module A:
```swift
class A {
    func a() {}
}
```

Generated file in a different module: 
```swift
@testable import A

class MockA: A, ... {
    ...
    override func a() {}
    ...
}
```

In the generated file the visibility of function `a` will be public because it was imported using `@testable`, but it's overridden with an internal function, therefore giving the error: `Overriding instance method must be as accessible as its enclosing type`

**Comment by TadeasKriz** [comment]
@Peter-Nagy Thank you for testing it. Could you push your changes to your fork, I'll clone it and try it locally. Thank you.

**Comment by TadeasKriz** [comment]
It seems I've broken the source so it couldn't compile. Let me fix that real quick.

**Comment by TadeasKriz** [comment]
@Peter-Nagy @sanoj00b Please try it again, now it'll compile as it should.


@Peter-Nagy I believe your example code would compile successfully. There's no such thing as "`a` will be public because it was imported using `@testable`". Visibility for `override` methods and properties is always bounded to the visibility of the child class. So if `MockA` is `internal`, then the `a` has to be at least `internal` to match the requirement set by class `A`. You could even make the `MockA` `private` and it would allow you to have `a` `private` too.

**Comment by sanoj00b** [comment]
@TadeasKriz  
```
open class Active {
	var service: Service = // Some singlton class object
}

//Generator mock file

public class ActiveStub: Active {
override var service: Service { **// error: Overriding property must be as accessible as its enclosing** type
        get {
            return DefaultValueRegistry.defaultValue(for: (Service).self)
        }
        set { }
    }
```

An error still coming. I think the generated mock file should add(Access Specifier) it automatically while compiling.

**Comment by TadeasKriz** [comment]
It should be working for you, I think the generator you're running is not build using the code from the branch. How do you run the generator?

**Comment by sanoj00b** [comment]
> It should be working for you, I think the generator you're running is not build using the code from the branch. How do you run the generator?

I am using that's why you fix/316 branch is mapping to latest "84ae4fe47e22dc489c3e0e4b8fb445e1f4270909" in Cartfile after running the Carthage update command. We run the generator using run script which has been added to build Phase in unit test target.

**Comment by TadeasKriz** [comment]
Carthage is only for the runtime, not the generator so you need to ensure the new generator is being used. Please show us the run script.

**Comment by sanoj00b** [comment]
> Carthage is only for the runtime, not the generator so you need to ensure the new generator is being used. Please show us the run script.

Below is the script we are using:

```
# Define output file. Change "$PROJECT_DIR/Tests" to your test's root source folder, if it's not the default name.
OUTPUT_FILE="$PROJECT_DIR/XXTests/Cuckoo/GeneratedMocks.swift"
echo "Generated Mocks File = $OUTPUT_FILE"

# Define input directory. Change "$PROJECT_DIR" to your project's root source folder, if it's not the default name.
INPUT_DIR="$PROJECT_DIR"
echo "Mocks Input Directory = $INPUT_DIR"

# Generate mock files, include as many input files as you'd like to create mocks for.
Carthage/Checkouts/Cuckoo/run generate --testable "$PROJECT_NAME" \
--output "${OUTPUT_FILE}" \
"$INPUT_DIR/XX/Services/Active" \
# ... and so forth

# After running once, locate `GeneratedMocks.swift` and drag it into your Xcode test target group.

```

**Comment by TadeasKriz** [comment]
@sanoj00b try adding `--clean` before the `--testable`.

**Comment by sanoj00b** [comment]
> @sanoj00b try adding `--clean` before the `--testable`.

Yey!! That worked now. Thank you so much for help and support.

**Comment by TadeasKriz** [comment]
@sanoj00b Awesome! Be sure to remove the `--clean` now so that it doesn't build the generator each time.

**Comment by sanoj00b** [comment]
@TadeasKriz Please let me know when will you merge fix/316 branch? Should I close this issue?

**Comment by MatyasKriz** [comment]
Hey, @sanoj00b. The branch has been merged into 1.3.1 and released. I'll close the issue for you, thanks for the report!
